### ISSUE TITLE
nats and generics

### ISSUE THREAD

**Issue by gedw99** [RangelReale/go-kit-typed#1]
Hey @RangelReale 

This work looks really interesting.

Just trying to work out the intent / rational for this. The generics allows the use of different transports more easily ?

**Comment by RangelReale** [comment]
no, the biggest thing is to avoid sending raw `interface{}` in endpoints, which requires casting.

Also making endpoints typed we can detect more errors at compile time instead of only at runtime, making it a little safer overral.

Most of the biggest changes are the ones shown on the root README.
